server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java

```diff
@@ -467,6 +472,42 @@ void postParse() {
             return newDocs;
         }
 
+        /**
+         * If the document is going to be index into a backing index of a data stream then
+         * check whether a timestamp field has been specified and that exactly one timestamp
+         * value has been specified.
+         */
+        private void validateDataStreamTimestampField() {
+            if (dataStream == null) {
+                // the index being index into is not part of a datas stream.
+                return;
+            }
+
+            int numStoredFields = 0;
+            int numPointFields = 0;
+            int numDocValuesFields = 0;
+            IndexableField[] fields = rootDoc().getFields(dataStream.getTimeStampField());
+            if (fields.length == 0) {
+                throw new IllegalArgumentException("required timestamp field is missing");
+            }
+
+            for (IndexableField field : fields) {
+                if (field.fieldType().pointDimensionCount() != 0) {
+                    numPointFields++;
+                }
+                if (field.fieldType().stored()) {
+                    numStoredFields++;
+                }
+                if (field.fieldType().docValuesType() != DocValuesType.NONE) {
+                    numDocValuesFields++;
+                }
+            }
+
+            if (numStoredFields > 1 || numPointFields > 1 || numDocValuesFields > 1) {
```

Comment:
It feels a little fragile to be checking the Lucene fields that the timestamp field produces. Sometimes field mappers decide to produce multiple Lucene fields given a single value in the _source. For example, mappers sometimes produce a special `_field_names` field to support exists queries.

Instead I think we could check that the timestamp is a 'singleton' during document parsing:
* We could add a boolean flag to `DateFieldMapper` like `isSingletonTimestamp`, based on whether its field name matches the datastream timestamp field.
* In `DateFieldMapper#parseCreateField` we would check + update a flag on `ParseContext` like `alreadyParsedTimestamp`. If it's already true, we throw an error.
* In `ParseContext#postParse`, we verify that `alreadyParsedTimestamp` is true.

I like this because it avoids the fragility of checking Lucene fields, but still correctly handle cases where data is copied into the field like `copy_to` and multi-fields.

Reply:
Thanks for this suggestion @jtibshirani. I also found counting of Lucene fields to be fragile and the `isSingletonTimestamp` and `alreadyParsedTimestamp` flags will make this check more robust. I will try and adjust the code.